packages/hardware-ledger/src/LedgerKeyAgent.ts

```diff
@@ -148,98 +165,207 @@ export class LedgerKeyAgent extends KeyAgentBase {
     this.#communicationType = serializableData.communicationType;
   }
 
-  static findKeyAgentByCommunicationTypeAndDevicePath(communicationType: CommunicationType, devicePath?: string) {
-    return this.deviceConnections?.find(
-      (connection) => connection.communicationType === communicationType && connection.devicePath === devicePath
+  private static async findConnectionByCommunicationTypeAndDevicePath(
+    communicationType: CommunicationType,
+    nodeHidDevicePath?: string,
+    device?: LedgerDevice
+  ): Promise<LedgerConnection | null> {
+    const matchingConnectionData = this.deviceConnections?.find(
+      (connection) =>
+        connection.communicationType === communicationType &&
+        connection.nodeHidDevicePath === nodeHidDevicePath &&
+        connection.device === device
     );
+    if (!matchingConnectionData) return null;
+
+    try {
+      await this.testConnection(matchingConnectionData.deviceConnection);
+    } catch (error) {
+      if (error instanceof errors.TransportError && error.message.includes(CARDANO_APP_CONNECTION_ERROR_MESSAGE)) {
+        this.deviceConnections = this.deviceConnections.filter(
+          (connectionData) => connectionData !== matchingConnectionData
+        );
+        return null;
+      }
+
+      throw error;
+    }
+
+    return matchingConnectionData.deviceConnection;
   }
 
   /**
    * @throws TransportError
    */
-  static async getHidDeviceList(communicationType: CommunicationType): Promise<string[]> {
+  private static async getHidDeviceList(communicationType: CommunicationType): Promise<string[]> {
     try {
-      return communicationType === CommunicationType.Node ? TransportNodeHid.list() : TransportWebHID.list();
+      return communicationType === CommunicationType.Node ? TransportNodeHid.list() : TransportWebUSB.list();
     } catch (error) {
       throw new errors.TransportError('Cannot fetch device list', error);
     }
   }
 
+  private static async closeTransport(transport: LedgerTransportType) {
+    try {
+      await transport.close();
+    } catch (error) {
+      // eslint-disable-next-line no-console
+      console.error('Failed to close the transport', transport, error);
+    }
+  }
+
+  private static attachDisconnectionCleanupHandler(transport: LedgerTransportType) {
+    const onDisconnect = () => {
+      transport.off('disconnect', onDisconnect);
+      this.deviceConnections = this.deviceConnections.filter(
+        ({ deviceConnection }) => deviceConnection.transport !== transport
+      );
+      void this.closeTransport(transport);
+    };
+    transport.on('disconnect', onDisconnect);
+  }
+
+  /**
+   * @throws TransportError
+   */
+  private static async openTransportForDevice({ communicationType, device }: OpenTransportForDeviceParams) {
+    let transport: LedgerTransportType;
+    try {
+      if (communicationType === CommunicationType.Node && device instanceof HID) {
```

Comment:
I think it's better to be defensive when using globals, as it might not exist in some environments. Same for USBDevice
```suggestion
      if (communicationType === CommunicationType.Node && typeof HID === 'function' && device instanceof HID) {
```

Reply:
Right, I overlooked the usage of the `USBDevice` here. With the `HID` we are safe because it is imported.